algorithmic_efficiency/workloads/librispeech_conformer/librispeech_pytorch/model.py

```diff
@@ -0,0 +1,424 @@
+    self.dropout1 = nn.Dropout(p=config.feed_forward_dropout_rate)
+    self.linear2 = nn.LazyLinear(out_features=config.encoder_dim, bias=True)
+    self.dropout2 = nn.Dropout(p=config.feed_forward_residual_dropout_rate)
+
+  def forward(self, inputs, padding_mask):
+    inputs = self.ln(inputs)
+    inputs = self.linear1(inputs)
+    inputs = F.silu(inputs)
+    inputs = self.dropout1(inputs)
+    inputs = inputs * padding_mask
+    inputs = self.linear2(inputs)
+    inputs = inputs * padding_mask
+    inputs = self.dropout2(inputs)
+
+    return inputs
+
+
+class AddPositionalEmbedding(nn.Module):
+
+  def __init__(self,
+               min_timescale: int = 1,
+               max_timescale: int = 10_000,
+               embedding_dim: int = 512):
+    super().__init__()
+    self.min_timescale = min_timescale
+    self.max_timescale = max_timescale
+    self.embedding_dim = embedding_dim
+    num_timescales = self.embedding_dim // 2
+    log_timescale_increment = math.log(
+        float(self.max_timescale) / float(self.min_timescale)) / (
+            num_timescales - 1)
+    inv_timescales = self.min_timescale * \
+        torch.exp(torch.arange(num_timescales, dtype=torch.float32)
+                  * -log_timescale_increment)
+    self.register_buffer("inv_timescales", inv_timescales[None, None, :])
+
+  def forward(self, seq_length):
+    position = torch.arange(
+        end=seq_length, dtype=torch.float32, device=self.inv_timescales.device)
+    scaled_time = position[None, :, None] * self.inv_timescales
+    signal = torch.cat([torch.sin(scaled_time), torch.cos(scaled_time)], dim=2)
+    if self.embedding_dim % 2:
+      signal = torch.cat(
+          [signal, torch.zeros(signal.shape[0], signal.shape[1], 1)], dim=2)
+    return signal
+
+
+class QueryScaler(nn.Module):
+
+  def __init__(self, dim):
+    super().__init__()
+    self.dim = dim
+    self.scale = nn.Parameter(torch.zeros(self.dim))
+
+  def forward(self, inputs):
+    r_softplus_0 = 1.442695041
+    scale = r_softplus_0 * F.softplus(self.scale)
+    return inputs * scale
+
+
+class MultiHeadedSelfAttention(nn.Module):
+
+  def __init__(self, config: ConformerConfig):
+    super().__init__()
+
+    self.config = config
+
+    self.ln = LayerNorm(dim=config.encoder_dim)
+    self.qs = QueryScaler(dim=config.encoder_dim)
+    self.self_attention = nn.MultiheadAttention(
+        embed_dim=config.encoder_dim,
+        num_heads=config.num_attention_heads,
+        dropout=config.attention_dropout_rate,
+        bias=True,
+        batch_first=True)
+    self.dropout = nn.Dropout(p=config.attention_residual_dropout_rate)
+
+  def forward(self, outputs, paddings):
+    outputs = self.ln(outputs)
+    outputs, _ = self.self_attention(
+        query=self.qs(outputs),
+        key=outputs,
+        value=outputs,
+        key_padding_mask=paddings,
+        need_weights=False,
+    )
+    outputs = self.dropout(outputs)
+    return outputs
+
+
+class BatchNorm(nn.Module):
```

Comment:
in order to make the scale init the same as jax, can we instead make another class that subclasses `nn.BatchNorm1d`, where the only thing it does is override `reset_parameters` and `forward`? looking at the [source](https://pytorch.org/docs/stable/_modules/torch/nn/modules/batchnorm.html#BatchNorm1d) I think that's all we need for it? `reset_parameters` would just init `weight` to 0, and `forward` would pass `(1 + self.weight)` to `F.batch_norm`, something like below:

```
    def reset_parameters(self) -> None:
        self.reset_running_stats()
        if self.affine:
            init.zeros_(self.weight)
            init.zeros_(self.bias)
```

```
    def forward(self, input: Tensor) -> Tensor:
        self._check_input_dim(input)

        # exponential_average_factor is set to self.momentum
        # (when it is available) only so that it gets updated
        # in ONNX graph when this node is exported to ONNX.
        if self.momentum is None:
            exponential_average_factor = 0.0
        else:
            exponential_average_factor = self.momentum

        if self.training and self.track_running_stats:
            # TODO: if statement only here to tell the jit to skip emitting this when it is None
            if self.num_batches_tracked is not None:  # type: ignore[has-type]
                self.num_batches_tracked.add_(1)  # type: ignore[has-type]
                if self.momentum is None:  # use cumulative moving average
                    exponential_average_factor = 1.0 / float(self.num_batches_tracked)
                else:  # use exponential moving average
                    exponential_average_factor = self.momentum

        r"""
        Decide whether the mini-batch stats should be used for normalization rather than the buffers.
        Mini-batch stats are used in training mode, and in eval mode when buffers are None.
        """
        if self.training:
            bn_training = True
        else:
            bn_training = (self.running_mean is None) and (self.running_var is None)

        r"""
        Buffers are only updated if they are to be tracked and we are in training mode. Thus they only need to be
        passed when the update should occur (i.e. in training mode when they are tracked), or when buffer stats are
        used for normalization (i.e. in eval mode when buffers are not None).
        """
        return F.batch_norm(
            input,
            # If buffers are not to be tracked, ensure that they won't be updated
            self.running_mean
            if not self.training or self.track_running_stats
            else None,
            self.running_var if not self.training or self.track_running_stats else None,
            (1.0 + self.weight),
            self.bias,
            bn_training,
            exponential_average_factor,
            self.eps,
        )
```

I'm not torch expert tho, so maybe I'm missing something.

Reply:
Nice, will do this!

Reply:
Done.